fix(ivy): let ngcc not consider deep imports as missing dependencies

JoostK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

NGCC would incorrectly skip packages that contain RxJS 5 style imports


## What is the new behavior?

RxJS 5 style imports (deep imports in general) will no longer cause a package to be skipped over by NGCC.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JoostK committed:
fix(ivy): let ngcc not consider deep imports as missing dependencies

This fixes an issue where packages would be skipped if they contained
e.g. RxJS 5 style imports such as
```
import { observeOn } from 'rxjs/operators/observeOn';
```

Given that no package.json file can be found at the imported path, the
dependency would be reported missing, causing the package to be skipped.

JoostK committed:
fixup! fix(ivy): let ngcc not consider deep imports as missing dependencies

This fixes an issue where packages would be skipped if they contained
e.g. RxJS 5 style imports such as
```
import { observeOn } from 'rxjs/operators/observeOn';
```

Given that no package.json file can be found at the imported path, the
dependency would be reported missing, causing the package to be skipped.

alxhub commented on packages/compiler-cli/src/ngcc/src/packages/dependency_host.ts:
```diff
@@ -51,11 +49,15 @@ export class DependencyHost {
+              // If the import cannot be resolved at all, register it as missing dependency.
+              const resolvedFile = this.tryResolve(from, importPath);
+              if (resolvedFile === null) {
+                missing.add(importPath);
```
Walk me through what happens if it is resolved to a file here? Then it's ignored? We should log a warning in this case.

JoostK replied:
I choose to ignore it, as we expect that this situation should not be troublesome. I will collect such imports and create a single warning from it.

JoostK committed:
fixup! fix(ivy): let ngcc not consider deep imports as missing dependencies

This fixes an issue where packages would be skipped if they contained
e.g. RxJS 5 style imports such as
```
import { observeOn } from 'rxjs/operators/observeOn';
```

Given that no package.json file can be found at the imported path, the
dependency would be reported missing, causing the package to be skipped.

petebacondarwin reviewed (approved):
I think that this is the right thing to do right now. In the future if we find that we have difficulties ordering dependencies based on deep imports then we can add to the strategy.

Thanks @JoostK - please check the minor comments I made.

JoostK committed:
fixup! fix(ivy): let ngcc not consider deep imports as missing dependencies

This fixes an issue where packages would be skipped if they contained
e.g. RxJS 5 style imports such as
```
import { observeOn } from 'rxjs/operators/observeOn';
```

Given that no package.json file can be found at the imported path, the
dependency would be reported missing, causing the package to be skipped.

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/dependency_host.ts:
```diff
@@ -17,15 +17,15 @@ export class DependencyHost {
-   * @param resolved A set that will have the resolved dependencies added to it.
+   * @param resolved A set that will have the absolute paths of resolved entry points added to it.
    * @param missing A set that will have the dependencies that could not be found added to it.
+   * @param deepImports A set that will have the dependencies that refer to deep imports added to
```
Done

JoostK commented on packages/compiler-cli/src/ngcc/test/packages/dependency_host_spec.ts:
```diff
@@ -27,76 +27,101 @@ describe('DependencyHost', () => {
+      expect(resolved.has('RESOLVED/path/to/x')).toBe(true);
       expect(missing.size).toBe(1);
-      expect(missing.has('missing'));
+      expect(missing.has('missing')).toBe(true);
```
Done!

JoostK commented on packages/compiler-cli/src/ngcc/test/packages/dependency_host_spec.ts:
```diff
@@ -27,76 +27,101 @@ describe('DependencyHost', () => {
+      spyOn(host, 'tryResolveEntryPoint')
+          .and.callFake(
+              (from: string, importPath: string) =>
+                  importPath === 'deep/import' ? null : `RESOLVED/${importPath}`);
```
Done

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/dependency_resolver.ts:
```diff
@@ -109,6 +110,13 @@ export class DependencyResolver {
+        const imports = Array.from(deepImports).map(i => `'${i}'`).join(', ');
+        console.warn(
+            `Entry point '${entryPoint.name}' contains deep imports into ${imports}. ` +
+            `This may cause compilation errors if the package that is deeply imported also needs to be processed by ngcc.`);
```
I rephrased it a bit:

```typescript

console.warn(
  `Entry point '${entryPoint.name}' contains deep imports into ${imports}. ` +
  `This is probably not a problem, but may cause the compilation of entry points to be out of order.`);
```

Still not sure if this is descriptive enough, or if we should actually go into more detail (or less?).

petebacondarwin replied:
That works for me.

JoostK commented on packages/compiler-cli/src/ngcc/test/packages/dependency_host_spec.ts:
```diff
@@ -27,76 +27,101 @@ describe('DependencyHost', () => {
+      const resolved = new Set();
+      const missing = new Set();
+      const deepImports = new Set();
+      host.computeDependencies('/external/deep-import.js', resolved, missing, deepImports);
```
Done
